core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java

```diff
@@ -286,6 +286,9 @@ private SqlLibraryOperators() {
               opBinding -> {
                 int precision = opBinding.collectOperandTypes().stream()
                     .mapToInt(RelDataType::getPrecision).sum();
+                // precision may be smaller than -1 when operands' precision are not be specified,
+                // which is not allowed.
+                precision = Math.max(precision, -1);
```

Comment:
In `concat(x, y, z)`, if the precision of `x`, `y`, `z` are specified (i.e., precision != `PRECISION_NOT_SPECIFIED`), the return type is the sum of them. However, if one operand's precision is `PRECISION_NOT_SPECIFIED`, the return type should also be `PRECISION_NOT_SPECIFIED`. 
E.g., for `concat('aaa', cast('bb' as varchar), 'ccc')`, operand's precisions are: [3, -1, 3]. The return type's precision (3-1+3=5) is not resonable.